File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        
        for(int i = 0; i < s.length(); i++){
            if(s.charAt(i) == '('){
                stack.add('(');
            }else if(s.charAt(i) == '{'){
                stack.add('{');
            }else if(s.charAt(i) == '['){
                stack.add('[');
            }else if(stack.size() == 0){
                return false;
            }else if(s.charAt(i) == ')' && stack.peek() == '('){
                stack.pop();
            } else if(s.charAt(i) == '}' && stack.peek() == '{'){
                stack.pop();
            }else if(s.charAt(i) == ']' && stack.peek() == '['){
                stack.pop();
            }else return false;
        }
        
        if(stack.size() != 0) return false;
        return true;
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        for(int i = 0; i < s.size(); i++){
            if(s[i] == '('){
                st.push(')');
            }else if(s[i] == '{'){
                st.push('}');
            }else if(s[i] == '['){
                st.push(']');
            }else{
                if(st.empty()) return false;
                
                if(st.top() == s[i]){
                    st.pop();
                }else return false;
            }
        }
        if(!st.empty()) return false;
        return true;
    }
};